viewing_party/party.py

```diff
@@ -1,23 +1,297 @@
+    # titles, but it gets the job done. 
+
+    movie_dict = {}
+    for movie in movie_list:
+        if movie['title'] not in movie_dict:
+            movie_dict[movie['title']] = movie 
+    
+    # make sure the fuction returns a list rather than 
+    # a dict.values() iterator
+    return list(movie_dict.values())
+
+
+def get_friends_unique_watched(user_data):
+    # takes user_data returns a list of movies watched by 
+    # friends but not watched by the user 
+
+    # all of these should be lists 
+    friends_watched = watched_by_friends(user_data)
+    my_watched = user_data.get('watched', [])
+    friends_not_unique = []
+
+    for movie in friends_watched:
+        if movie not in my_watched:
+            friends_not_unique.append(movie)
+
+    # remove duplicates from friends_not_unique
+    friends_unique = remove_duplicates_by_title(friends_not_unique)
+
+    return friends_unique
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    # return a list of recommendations from friends
+    # that are available through subscribed streaming services
+
+    subscriptions = user_data['subscriptions']
+    movies_from_friends = remove_duplicates_by_title(watched_by_friends(user_data))
+    my_watched_titles = [movie['title'] for movie in user_data['watched']]
+
+    recs = []
+    for movie in movies_from_friends:
+        if movie['title'] not in my_watched_titles:
+            if movie['host'] in subscriptions:
+                recs.append(movie)
+    return recs 
+        
+            
+
+
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def most_frequent_genre(user_data):
+    # get the most frequent genre from user_data
+
+    genre_freq_table = {}
+    for movie in user_data['watched']:
+        if movie['genre'] in genre_freq_table:
+            genre_freq_table[movie['genre']] += 1
+        else:
+            genre_freq_table[movie['genre']] = 1
+
+    top_genre = ""
+    top_score = 0
+    if len(genre_freq_table) == 0:
+        return None 
+    for genre, score in genre_freq_table.items():
+        if score > top_score:
+            top_score = score
+            top_genre = genre
+    
+    return top_genre
+
+def get_new_rec_by_genre(user_data):
+    # get a new recommendation that matches the 
+    # favorite genre and hasn't been seen. 
+
+    movies_from_friends = remove_duplicates_by_title(watched_by_friends(user_data))
+    favorite_genre = most_frequent_genre(user_data)
+
+    # 1. The user has not watched it
+    # 2. At least one of the user's friends has watched
+    # 3. The "genre" of the movie is the same as the user's most frequent genre
+
+    movie_recs = []
+    for movie in movies_from_friends:
+        if (movie['genre'] == favorite_genre) and (movie not in user_data['watched']):

```

Comment:
Rather than filtering the movies against the user's watched list here, we could start with `get_friends_unique_watched` to generate `movies_from_friends`.